crates/fluvio-protocol/src/core/wasm_bytes.rs

```diff
@@ -0,0 +1,87 @@
+use std::io::Error;
+use std::io::ErrorKind;
+
+use bytes::{Bytes, BufMut};
+
+use crate::{Encoder, Version};
+
+/// Represnts a SmartModule WASM File bytes.
+///
+/// Provides a `Encoder` implementation optimized for WASM files used in
+/// SmartModules.
+pub struct WasmBytes(Vec<u8>);
+
+impl From<Vec<u8>> for WasmBytes {
+    fn from(bytes: Vec<u8>) -> Self {
+        Self(bytes)
+    }
+}
+
+impl Encoder for WasmBytes {
+    fn write_size(&self, version: Version) -> usize {
+        self.0
+            .iter()
+            .fold(4, |sum, val| sum + val.write_size(version))
+    }
+
+    fn encode<T>(&self, dest: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: BufMut,
+    {
+        let remaining = dest.remaining_mut();
+        let expected = self.write_size(version);
+
+        if remaining < expected {
+            return Err(Error::new(
+                ErrorKind::UnexpectedEof,
+                format!(
+                    "Not enough capacity for WasmBytes. Expected: {}, Remaining: {}",
+                    expected, remaining
+                ),
+            ));
+        }
+
+        dest.put_u32(self.0.len() as u32);
+        dest.put_slice(self.0.as_slice());
```

Comment:
Given that we acknowledge the contents of the underlying `Vec`, first puts the length of the `Vec` and then the whole contents as a slice instead of encoding -> pushing to dest.